src/js/view/controls/components/menu/menu.js

```diff
@@ -0,0 +1,499 @@
+        }
+        this.visible = true;
+        this.el.setAttribute('aria-expanded', 'true');        
+    }
+    close(evt) {
+        // if visible, return
+        if (!this.visible) {
+            return;
+        }
+        if (this.isSubmenu) {
+            this.el.classList.remove('jw-settings-submenu-active');
+            this.categoryButton.element().setAttribute('aria-checked', 'false');
+            this.mainMenu.openMenus = this.mainMenu.openMenus.filter(name => name !== this.name);
+            if (!this.mainMenu.openMenus.length && this.mainMenu.visible) {
+                this.mainMenu.close(evt);
+            }
+        } else {
+            this.el.parentNode.classList.remove('jw-settings-open');
+            this.trigger('menuVisibility', { visible: false, evt });
+            document.removeEventListener('click', this.onDocumentClick);
+            if (this.mainMenu.openMenus.length) {
+                this.mainMenu.closeAll();
+            }
+        }
+        this.visible = false;
+        this.el.setAttribute('aria-expanded', 'false');  
+    }
+    closeAll() {
+        this.openMenus.forEach(menuName => {
+            this.children[menuName].close();
+        });
+    }
+    toggle(evt) {
+        if (this.visible) {
+            this.close(evt);
+        } else {
+            this.open(evt);
+        }
+    }
+    onDocumentClick(e) {
+        if (!/jw-(settings|video|nextup-close|sharing-link|share-item)/.test(e.target.className)) {
+            this.close();
+        }
+    }
+    destroy() {
+        document.removeEventListener('click', this.onDocumentClick);
+        if (this.isSubmenu) {
+            this.parentMenu.topbar.removeChild(this.categoryButton.element());
+            this.categoryButton.ui.destroy();
+        } else {
+            this.ui.destroy();
+        }
+        this.visible = false;
+        if (this.el.parentNode) {
+            this.el.parentNode.removeChild(this.el);
+        }
+        
+        Object.keys(this.children).map(menuName => {
+            this.children[menuName].destroy();
+        });
+    }
+}
+
+const addGlobalMenuKeyListener = (settingsMenu) => {
+    const closeButton = settingsMenu.closeButton;
+    const settingsMenuElement = settingsMenu.el;
+    const ui = new UI(settingsMenuElement).on('keydown', function(evt) {
+        const { sourceEvent, target } = evt;
+        const next = nextSibling(target);
+        const prev = previousSibling(target);
+        const key = sourceEvent.key.replace(/(Arrow|ape)/, '');
+        const onLeft = (isTab) => {
+            if (prev) {
+                if (!isTab) {
+                    prev.focus();
+                }
+            } else {
+                settingsMenu.close(evt);
+            }
+        };
+        const onRight = () => {
+            if (next && closeButton.element() && target !== closeButton.element()) {
+                next.focus();
+            }
+        };
+        let childMenu;
+        switch (key) {
```

Comment:
i'm assuming this code is straight ported from the old given the patterns it matches in the code above

Reply:
in that vein, my comments there apply here as well regarding style etc, but i wouldn't go too far in rewriting this stuff, esp if the structural change is more the effort of this pr

Reply:
it is almost entirely ported